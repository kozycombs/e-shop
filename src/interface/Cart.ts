export interface CartProduct {
  productId: number;
  quantity: number;
}

export interface Cart {
  id: number;
  userId: number;
  date: string;
  products: CartProduct[];
}

export interface CartResponse {
  id: number;
}
