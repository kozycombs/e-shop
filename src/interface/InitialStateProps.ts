export interface InitialStateProps<T> {
  data: T;
  error: string;
}
