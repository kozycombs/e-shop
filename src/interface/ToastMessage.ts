export interface ToastMessage {
  message: string;
  severity: "success" | "error";
}
