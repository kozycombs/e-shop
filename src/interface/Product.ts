import { Rating } from "interface/Rating";

export interface Product {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: Rating;
}
