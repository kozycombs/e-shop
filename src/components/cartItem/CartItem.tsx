import { FC } from "react";
import { CartProduct } from "../../interface/Cart";
import { Product } from "../../interface/Product";
import { Box, Divider, Grid, Typography } from "@mui/material";
import { Link } from "react-router-dom";
import { APP_URL } from "../../constants";

interface CartItemProps {
  cartItem: CartProduct;
  product: Product;
}

const CartItem: FC<CartItemProps> = ({ cartItem, product }) => {
  return (
    <Grid
      container
      spacing={{ xs: 4, md: 3 }}
      rowSpacing={2}
      columns={{ xs: 1, sm: 5 }}
      sx={{ mb: 5 }}
    >
      <Grid item xs={1} sm={1}>
        <Link to={`${APP_URL.PRODUCT}/${product.id}`}>
          <Box
            component="img"
            sx={{
              height: "auto",
              width: "100%",
              maxHeight: { xs: 100 },
              objectFit: "contain",
            }}
            alt={product?.title}
            src={product?.image}
          />
        </Link>
      </Grid>
      <Grid item xs={1} sm={3}>
        <Link to={`${APP_URL.PRODUCT}/${product.id}`}>
          <Typography component="p" sx={{ fontSize: 18, fontWeight: 500 }}>
            {product.title}
          </Typography>
          <Typography
            component="p"
            sx={{ fontSize: 14, fontWeight: 400, mb: 4 }}
          >
            quantity: {cartItem.quantity}
          </Typography>
        </Link>
      </Grid>

      <Grid item xs={1} sm={1}>
        <Box>
          <Typography
            component="p"
            sx={{ fontSize: 18, fontWeight: 400, mb: 4 }}
          >
            &pound;{product?.price * cartItem.quantity}
          </Typography>
        </Box>
      </Grid>
      <Divider sx={{ mb: 4 }} />
    </Grid>
  );
};

export default CartItem;
