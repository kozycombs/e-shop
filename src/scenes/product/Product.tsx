import { FC, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {
  Alert,
  Box,
  Button,
  Divider,
  FormControl,
  Grid,
  MenuItem,
  Rating,
  Select,
  SelectChangeEvent,
  Snackbar,
  Typography,
} from "@mui/material";
import { fetchProduct, productsSelector } from "store/productSlice";
import { Product as ProductInterface } from "interface/Product";
import Spinner from "components/spinner/Spinner";
import { addToCart, cartSelector, updateCartProduct } from "store/cartSlice";
import { AppDispatch } from "store/index";
import { ToastMessage } from "interface/ToastMessage";
import { PRODUCT_QUATITY_COUNT } from "constants/index";
import ErrorBanner from "components/errorBanner/ErrorBanner";

const Product: FC = () => {
  const { id } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [loading, setLoading] = useState(false);
  const [product, setProduct] = useState<ProductInterface>();
  const [quantity, setQuantity] = useState(1);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [error, setError] = useState("");
  const userId = 1; // This is a dummy user id, since there is no logged in functionality.
  const vertical = "top";
  const horizontal = "right";

  useEffect(() => {
    const getProduct = async () => {
      try {
        const result = await dispatch(fetchProduct(Number(id))).unwrap();
        setProduct(result);
        setLoading(false);
      } catch (error) {
        setError("Failed to load our product. Please try again later");
        setLoading(false);
      }
    };

    // If product list is empty get product detail from API
    // otherwiser get product detail from cached product list.

    if (products.length <= 0) {
      setLoading(true);
      getProduct();
    } else {
      setLoading(true);
      const product =
        products.find((currentProduct) => currentProduct.id === Number(id)) ||
        null;
      if (product) {
        setProduct(product);
      }
      setLoading(false);
    }
  }, [id, products, dispatch]);

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const handleQuantityChange = (event: SelectChangeEvent) => {
    setQuantity(Number(event.target.value));
  };

  const handleAddToCart = async () => {
    if (!product) {
      return;
    }

    try {
      setLoading(true);
      const date = new Date().toISOString();

      // If cart is null add to cart else update with cart id.
      if (!cart) {
        const result = await dispatch(
          addToCart({
            userId,
            date,
            products: [{ productId: product?.id, quantity }],
          })
        ).unwrap();
        if (result.id) {
          setToastMessage({
            message: "Product added to cart successfully",
            severity: "success",
          });
        } else {
          setToastMessage({
            message: "Failed to add product to cart",
            severity: "error",
          });
        }
      } else {
        const productFoundIndex = cart.products.findIndex(
          (currentProduct) => currentProduct.productId === product.id
        );

        if (productFoundIndex > -1) {
          const updatedProduct = {
            ...cart.products[productFoundIndex],
            quantity: cart.products[productFoundIndex].quantity + quantity,
          };
          const otherProducts = cart.products.filter(
            (currentProduct) => currentProduct.productId !== product.id
          );
          const products = [...otherProducts, updatedProduct];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        } else {
          const products = [
            ...cart.products,
            { productId: product?.id, quantity },
          ];
          const result = await dispatch(
            updateCartProduct({ cartId: cart.id, userId, date, products })
          ).unwrap();
          if (result.id) {
            setToastMessage({
              message: "Product added to cart successfully",
              severity: "success",
            });
            setQuantity(1);
          } else {
            setToastMessage({
              message: "Failed to add product to cart",
              severity: "error",
            });
          }
        }
      }
      setLoading(false);
    } catch (error) {
      setError("Failed to add product to cart. Please try again later");
      setLoading(false);
    }
  };

  return (
    <Box sx={{ pt: 4 }}>
      {loading && (
        <Box className="flex justify-center">
          <Spinner />
        </Box>
      )}

      {error.length > 0 && <ErrorBanner message={error} />}

      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>

      <Grid container spacing={{ xs: 4, md: 3 }} columns={{ xs: 1, sm: 4 }}>
        <Grid item xs={1} sm={2} sx={{ marginBottom: 10 }}>
          <Box
            component="img"
            sx={{
              height: "auto",
              width: "100%",
              maxHeight: { xs: 400 },
              objectFit: "contain",
            }}
            alt={product?.title}
            src={product?.image}
          />
        </Grid>
        <Grid item xs={1} sm={2}>
          <Box></Box>
          <Typography
            variant="h1"
            sx={{ mb: 1, fontWeight: "400", fontSize: 24 }}
          >
            {product?.title}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "end", mb: 2 }}>
            <Typography
              component="span"
              sx={{
                mr: 1,
                fontSize: 16,
              }}
            >
              {product?.rating.rate}
            </Typography>
            <Rating
              name="read-only"
              value={Number(product?.rating.rate)}
              readOnly
            />
            <Typography
              component="span"
              sx={{
                ml: 1,
                mt: 1,
                fontSize: 16,
                textDecoration: "underline",
              }}
            >
              {product?.rating.count} reviews
            </Typography>
          </Box>
          <Typography
            component="p"
            sx={{ fontSize: 24, fontWeight: 400, mb: 4 }}
          >
            &pound;{Number(product?.price).toFixed(2)}
          </Typography>
          <Box sx={{ display: "flex", alignItems: "center", mb: 4 }}>
            <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
              <Select
                id="quantity"
                data-test="quantity"
                value={`${quantity}`}
                onChange={handleQuantityChange}
              >
                {Array(PRODUCT_QUATITY_COUNT)
                  .fill("_")
                  .map((_, index) => (
                    <MenuItem key={index} value={index + 1}>
                      {index + 1}
                    </MenuItem>
                  ))}
              </Select>
            </FormControl>
            <Button
              variant="contained"
              size="medium"
              disableElevation
              onClick={handleAddToCart}
              sx={{ textTransform: "none" }}
            >
              Add to Cart
            </Button>
          </Box>
          <Divider />
          <Typography
            component="p"
            sx={{
              mb: 2,
              mt: 4,
              fontSize: 16,
            }}
          >
            {product?.description}
          </Typography>
        </Grid>
      </Grid>
    </Box>
  );
};

export default Product;
