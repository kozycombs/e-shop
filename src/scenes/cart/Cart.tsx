import { FC, useState } from "react";
import {
  Alert,
  Box,
  Button,
  Card,
  CardContent,
  Divider,
  Grid,
  Snackbar,
  Typography,
} from "@mui/material";
import { useDispatch, useSelector } from "react-redux";
import { productsSelector } from "../../store/productSlice";
import { cartSelector, resetCart } from "../../store/cartSlice";
import CartItem from "../../components/cartItem/CartItem";
import { ToastMessage } from "../../interface/ToastMessage";
import { Link } from "react-router-dom";
import { APP_URL } from "../../constants";

const Cart: FC = () => {
  const dispatch = useDispatch();
  const { data: products } = useSelector(productsSelector);
  const { data: cart } = useSelector(cartSelector);
  const [toastMessage, setToastMessage] = useState<ToastMessage>({
    message: "",
    severity: "success",
  });
  const [orderConfirmed, setOrderConfirmed] = useState(false);
  const vertical = "top";
  const horizontal = "right";

  const resetToastMessage = () => {
    setToastMessage({ message: "", severity: "success" });
  };

  const getCartTotal = () => {
    return cart?.products.reduce((accumulator, cartItem) => {
      const cartProduct = products.find(
        (product) => product.id === cartItem.productId
      );
      if (!cartProduct) {
        return 0;
      }
      return (accumulator += cartProduct.price * cartItem.quantity);
    }, 0);
  };

  const handlePlaceOrder = () => {
    setOrderConfirmed((prev) => !prev);
    dispatch(resetCart());
  };

  if (orderConfirmed) {
    return (
      <Box sx={{ pt: 4 }}>
        <Typography variant="h1" sx={{ fontSize: 30, mb: 2 }}>
          Order Confirmation
        </Typography>
        <Divider sx={{ mb: 4 }} />
        <Typography>Your order was placed successfully.</Typography>
      </Box>
    );
  }

  return (
    <Box sx={{ pt: 4 }}>
      <Snackbar
        anchorOrigin={{ vertical, horizontal }}
        open={toastMessage?.message.length > 0}
        autoHideDuration={2000}
        onClose={resetToastMessage}
        key={vertical + horizontal}
      >
        <Alert
          onClose={resetToastMessage}
          severity={toastMessage?.severity}
          variant="filled"
          sx={{ width: "100%" }}
        >
          {toastMessage?.message}
        </Alert>
      </Snackbar>
      <Typography variant="h1" sx={{ fontSize: 30, mb: 2 }}>
        Your Basket
      </Typography>
      <Divider sx={{ mb: 4 }} />
      {!cart || cart?.products.length <= 0 ? (
        <Box sx={{ display: "flex", justifyContent: "space-around" }}>
          <Typography>
            Your basket is empty. View our{" "}
            <Link to={APP_URL.HOME} className="underline hover:no-underline">
              range of products
            </Link>
            .
          </Typography>
        </Box>
      ) : (
        <Grid container spacing={{ xs: 1, md: 4 }} columns={{ xs: 1, sm: 6 }}>
          <Grid item xs={1} sm={4}>
            {cart?.products.map((cartItem, index) => {
              const cartProduct = products.find(
                (product) => product.id === cartItem.productId
              );
              if (!cartProduct) {
                return "";
              }
              return (
                <CartItem
                  key={index}
                  cartItem={cartItem}
                  product={cartProduct}
                />
              );
            })}
          </Grid>
          <Grid item xs={1} sm={2}>
            <Card variant="outlined">
              <CardContent>
                <Typography variant="h2" sx={{ fontSize: 24, mb: 2 }}>
                  Summary
                </Typography>
                <Divider sx={{ mb: 4 }} />
                <Box
                  sx={{
                    display: "flex",
                    justifyContent: "space-between",
                    alignItems: "center",
                    mb: 1,
                  }}
                >
                  <Typography
                    component="p"
                    sx={{ fontSize: 18, fontWeight: 500, mb: 4 }}
                  >
                    Total
                  </Typography>
                  <Typography
                    component="p"
                    sx={{ fontSize: 18, fontWeight: 500, mb: 4 }}
                  >
                    &pound;{getCartTotal()}
                  </Typography>
                </Box>
                <Button
                  variant="contained"
                  size="medium"
                  color="success"
                  onClick={handlePlaceOrder}
                  sx={{ ml: "auto", display: "block" }}
                >
                  Place you order
                </Button>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      )}
    </Box>
  );
};

export default Cart;
