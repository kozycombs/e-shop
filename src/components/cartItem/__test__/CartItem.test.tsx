import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import CartItem from "../CartItem";
import { Product } from "../../../interface/Product";
import { CartProduct } from "../../../interface/Cart";

describe("<CartItem /> component", () => {
  const product: Product = {
    id: 16,
    title:
      "Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket",
    price: 29.95,
    description:
      "100% POLYURETHANE(shell) 100% POLYESTER(lining) 75% POLYESTER 25% COTTON (SWEATER), Faux leather material for style and comfort / 2 pockets of front, 2-For-One Hooded denim style faux leather jacket, Button detail on waist / Detail stitching at sides, HAND WASH ONLY / DO NOT BLEACH / LINE DRY / DO NOT IRON",
    category: "women's clothing",
    image: "https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg",
    rating: {
      rate: 2.9,
      count: 340,
    },
  };

  const cartItem: CartProduct = {
    productId: 16,
    quantity: 3,
  };

  it("should render the component", () => {
    const { asFragment } = render(
      <BrowserRouter>
        <CartItem product={product} cartItem={cartItem} isLast />
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
