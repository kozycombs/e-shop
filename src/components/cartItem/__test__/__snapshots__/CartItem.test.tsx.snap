// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> component should render the component 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-spacing-md-3 css-1njvgv6-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-1 css-w9nz8z-MuiGrid-root"
    >
      <a
        href="/product/16"
      >
        <img
          alt="Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket"
          class="MuiBox-root css-10514to"
          src="https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg"
        />
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-3 css-1rn35ll-MuiGrid-root"
    >
      <a
        href="/product/16"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1xh9tsm-MuiTypography-root"
        >
          Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-15okrha-MuiTypography-root"
        >
          quantity: 3
        </p>
      </a>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 MuiGrid-grid-sm-1 css-w9nz8z-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-0"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-jcjl1m-MuiTypography-root"
        >
          Â£89.85
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
